### PR TITLE
Remove port range requests

### PR DESCRIPTION
**Pull request by CjS77**
The approach of scanning ports in a range to find ephemeral ports to make connections with was
causing problems on two counts:
1. Rust runs tests in parallel, and scanning from 10000 for each test meant that there was
  a race condition when two tests check a port, see that it's free and then one gets it and the other
  does not.
2. Race conditions between `TcpListener::bind` and `zmq::connect`. Basically to see if a port is free
   we'd bind to it. If that was succesful, drop the bind and then give amq the port. If the OS didn't
   release the port in time, the zmq::connect call would fail.

We take a new approach to this here. Firstly, outbound connections must have their port specified
(This is typically done in config files). Inbound connections can have a random port selected
by using zero as the port, i.e. `0.0.0.0:0`. This delegates port selection to the OS and presumably
it does a better job of handling these race conditions.

After the connection is made, you can see the actual port that was used by calling
`conn.get_connected_address()`

This hopefully addresses the test flakiness and presents a more robust approach to port selection.
As a result all the port range and port selection logic has been removed as well.

### CODE REVIEW

File: comms/src/connection/connection.rs [hunk 1]
```diff
@@ -340,9 +355,32 @@ impl EstablishedConnection {
...
+    let addr = match socket.get_last_endpoint() {
+        Ok(v) => v.unwrap(),
+        Err(_) => return None,
+    };
+    let parts = &addr.split("//").collect::<Vec<&str>>();
+    if parts.len() < 2 || parts[0] != "tcp:" {
+        return None;
+    }
+    let addr = parts[1];
+    SocketAddress::from_str(&addr).ok()
```
Comment (sdbondi) [inline comms/src/connection/connection.rs]: For another PR, but it would be useful to have a NetAddress returned for onion and i2p addresses

File: comms/tests/connection/peer_connection.rs [hunk 2]
```diff
@@ -217,7 +214,7 @@ fn connection_wait_connect_failed() {
 
 #[test]
 fn connection_pause_resume() {
-    let addr = find_available_tcp_net_address("127.0.0.1").unwrap();
+    let addr = NetAddress::from_str("127.0.0.1:9873").unwrap();
```
Comment (sdbondi) [inline comms/tests/connection/peer_connection.rs]: Think a hard-coded port number could cause issues, if we rely on devs to use a unique port for their tests - the reason that the connection manager port allocations had a race condition was because there were multiple instances of connection manager, the test helper `find_available_tcp_net_address` port tracking is static and guarantees that you'll be able to "book" a port for the test. I don't think it was the cause of any flakiness.

File: comms/src/connection/connection.rs (hunk 1 above)
Comment (CjS77) [reply comms/src/connection/connection.rs]: Agreed. I'm sure there are edge cases to catch in this function

File: comms/tests/connection/peer_connection.rs (hunk 2 above)
Comment (CjS77) [reply comms/tests/connection/peer_connection.rs]: I'm worried that calling `bind` to see if a port is free and then giving it to zmq was also a source of flakiness -- if the OS holds onto the port for any length of time before zmq wants to connect...
